src/lib.rs

```diff
@@ -50,36 +58,95 @@ pub fn run(args: Args) {
             .expect(&format!("Unknown capture name: `{}`", capture_name))
     });
 
-    enumerate_project_files(&*supported_language)
-        .par_iter()
-        .for_each(|project_file_dir_entry| {
-            let mut printer = grep::printer::Standard::new_no_color(io::stdout());
-            let path = project_file_dir_entry.path();
-
-            let matcher = TreeSitterMatcher::new(
-                &query,
-                capture_index,
-                language,
-                args.filter.clone(),
-                args.filter_arg.clone(),
-            );
+    get_project_file_walker(&*supported_language)
+        .into_parallel_iterator()
+        .for_each({
+            |project_file_dir_entry| {
+                let mut printer = grep::printer::Standard::new_no_color(io::stdout());
+                let path = project_file_dir_entry.path();
+
+                let matcher = TreeSitterMatcher::new(
+                    &query,
+                    capture_index,
+                    language,
+                    args.filter.clone(),
+                    args.filter_arg.clone(),
+                );
+
+                SearcherBuilder::new()
+                    .multi_line(true)
+                    .build()
+                    .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
+                    .unwrap();
+            }
+        });
+}
 
-            SearcherBuilder::new()
-                .multi_line(true)
-                .build()
-                .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
-                .unwrap();
-        })
+trait IntoParallelIterator {
+    fn into_parallel_iterator(self) -> IterBridge<WalkParallelIterator>;
+}
+
+impl IntoParallelIterator for WalkParallel {
+    fn into_parallel_iterator(self) -> IterBridge<WalkParallelIterator> {
+        WalkParallelIterator::new(self).par_bridge()
+    }
+}
+
+struct WalkParallelIterator {
+    receiver_iterator: <Receiver<DirEntry> as IntoIterator>::IntoIter,
+    _handle: JoinHandle<()>,
+}
+
+impl WalkParallelIterator {
+    pub fn new(walk_parallel: WalkParallel) -> Self {
+        let (sender, receiver) = mpsc::channel::<DirEntry>();
+        let handle = thread::spawn(move || {
+            walk_parallel.run(move || {
+                Box::new({
+                    let sender = sender.clone();
+                    move |entry| {
+                        let entry = match entry {
+                            Err(_) => return WalkState::Continue,
+                            Ok(entry) => entry,
+                        };
+                        if !entry.metadata().unwrap().is_file() {
+                            return WalkState::Continue;
+                        }
+                        sender.send(entry).unwrap();
+                        WalkState::Continue
+                    }
+                })
+            });
+        });
+        Self {
+            receiver_iterator: receiver.into_iter(),
```

Comment:
Doesn't really matter, but why convert this into an `iter` here rather than in `next` below? Because then you'd have to consume it?

Reply:
Ya I guess (I was confused `Receiver` implements `IntoIterator` not `Iterator` directly) it seemed like what this struct needed as its internal state was the actual "inner `Iterator`" (which its `Iterator` implementation can just literally delegate to, interesting how unless there's a reason to hold onto the thread join handle reference (which I'm not sure - do bad things happen if you don't keep a thread join handle in scope?) this struct boils down to a basically non-existent wrapper around that inner `Iterator`)